src/java/com/rapleaf/jack/queries/GenericQuery.java

```diff
@@ -0,0 +1,147 @@
+package com.rapleaf.jack.queries;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import com.rapleaf.jack.BaseDatabaseConnection;
+import com.rapleaf.jack.ModelField;
+import com.rapleaf.jack.ModelWithId;
+
+public class GenericQuery {
+
+  private final BaseDatabaseConnection dbConnection;
+  private Class<? extends ModelWithId> mainModel;
+  private final List<JoinCondition> joinConditions;
+  private final List<WhereConstraint> whereConstraints;
+  private final List<OrderCriterion> orderCriteria;
+  private final Set<ModelField> selectedModelFields;
+  private Optional<LimitCriterion> limitCriteria;
+
+  private GenericQuery(BaseDatabaseConnection dbConnection) {
+    this.dbConnection = dbConnection;
+    this.mainModel = null;
+    this.joinConditions = Lists.newArrayList();
+    this.whereConstraints = Lists.newArrayList();
+    this.orderCriteria = Lists.newArrayList();
+    this.selectedModelFields = Sets.newHashSet();
+    this.limitCriteria = Optional.absent();
+  }
+
+  public static GenericQuery create(BaseDatabaseConnection dbConnection) {
+    return new GenericQuery(dbConnection);
+  }
+
+  public GenericQueryBuilder from(Class<? extends ModelWithId> model) {
+    this.mainModel = model;
+    return new GenericQueryBuilder(dbConnection, this);
+  }
+
+  List<WhereConstraint> getWhereConstraints() {
+    return whereConstraints;
+  }
+
+  Set<ModelField> getSelectedModelFields() {
+    return selectedModelFields;
+  }
+
+  void addJoinCondition(JoinCondition joinCondition) {
+    this.joinConditions.add(joinCondition);
+  }
+
+  void addWhereCondition(WhereConstraint whereConstraint) {
+    this.whereConstraints.add(whereConstraint);
+  }
+
+  void addOrderCondition(OrderCriterion orderCriterion) {
+    this.orderCriteria.add(orderCriterion);
+  }
+
+  void addLimitCondition(LimitCriterion limitCriterion) {
+    this.limitCriteria = Optional.of(limitCriterion);
+  }
+
+  void addSelectedModelField(ModelField modelField) {
+    this.selectedModelFields.add(modelField);
+  }
+
+  boolean isOrderedQuery() {
+    return !orderCriteria.isEmpty();
+  }
+
+  String getSqlStatement() {
+    StringBuilder statement = new StringBuilder();
+
+    statement.append(getSelectClause())
+        .append(getJoinClause())
+        .append(getWhereClause());
+
+    if (isOrderedQuery()) {
+      statement.append(getOrderClause());
+      statement.append(getLimitClause());
```

Comment:
I will add support for `limit` without order. But the result may be non-deterministic. That's my concern. Currently working on self-join, group by and custom table alias (that's a must for self-join). The current design of self-join syntax is like this:

```java
.from(Employee.class)
.leftJoin(Employee.class).as("manager").on(Employee.NAME.of("manager"), User.MANAGER_NAME)
```
So in this table, an employee can be the manager of another employee. The `Employee` table is aliased to `manager` in the join, and the join condition is `employee.manager_name = manager.name`.

When selecting a field from a self-joined table, the user needs to specify which table he / she wants (if that table is aliased).

```java
.select(Employee.NAME.of("manager"), Employee.NAME)
```
The above selection would give the name of the manager and the one he / she manages. The first `ModelField` is from the aliased `employee` table. The second one is from the non-aliased `employee` table.

The chained `.leftJoin().as().on()` structure is not necessary though. All these parameters can be passed into one `leftJoin` call as it is now. Probably the former is more readable.